protocol/testutil/app/gov.go

```diff
@@ -0,0 +1,127 @@
+package app
+
+import (
+	"bytes"
+	"testing"
+	"time"
+
+	abcitypes "github.com/cometbft/cometbft/abci/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypesv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	"github.com/dydxprotocol/v4-chain/protocol/testutil/constants"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	TestMetadata               = "test metadata"
+	TestTitle                  = "test title"
+	TestSummary                = "test summary"
+	TestSubmitProposalTxHeight = uint32(2)
+	TestProposalTallyHeight    = uint32(10)
+)
+
+var (
+	TestVotingPeriod = 1 * time.Minute
+)
+
+// SubmitAndTallyProposal simulates the following:
+//   - A proposal with the given messages is submitted. Check proposal submission succeeds or fails as expected.
+//   - If proposal successfully submitted:
+//     -- All validators vote for the proposal.
+//     -- The proposal is tallied after voting period ends.
+func SubmitAndTallyProposal(
+	t *testing.T,
+	ctx sdk.Context,
+	tApp *TestApp,
+	messages []sdk.Msg,
+	expectSubmitProposalFails bool,
+	expectedProposalStatus govtypesv1.ProposalStatus,
+) sdk.Context {
+	// Create a MsgSubmitProposal
+	msgSubmitProposal, err := govtypesv1.NewMsgSubmitProposal(
+		messages,
+		sdk.Coins{sdk.NewInt64Coin(constants.TestNativeTokenDenom, 10_000_000)},
+		constants.AliceAccAddress.String(),
+		TestMetadata,
+		TestTitle,
+		TestSummary,
+	)
+	require.NoError(t, err)
+
+	// Create a signed transaction with MsgSubmitProposal
+	submitProposalCheckTx := MustMakeCheckTxWithPrivKeySupplier(
+		ctx,
+		tApp.App,
+		MustMakeCheckTxOptions{
+			AccAddressForSigning: constants.AliceAccAddress.String(),
+			Gas:                  constants.TestGasLimit,
+			FeeAmt:               constants.TestFeeCoins_5Cents,
```

Comment:
for my understanding: does CheckTx fail if we don't pass in a fee?

Reply:
Yes